websqrl/workers/compile.worker.ts

```diff
@@ -191,38 +193,64 @@ async function runEvent(event: EventData, requestFeatures: readonly string[]) {
     features,
   };
 }
+
 function debounceCompile() {
   if (!compilePromise) {
     clearTimeout(compileTimeout);
     compileTimeout = setTimeout(triggerCompile, COMPILE_DEBOUNCE_MS);
   }
 }
 
-self.addEventListener("message", (event) => {
+let eventDomain: EventDomain;
+let fetchFeatures: readonly string[] = [];
+let eventProcessTimer: NodeJS.Timer | undefined;
+
+/** Queues up the next event for processing */
+const queueNextEvent = () => {
+  eventDomain.getNextEventDelay().then((eventDelay) => {
+    eventProcessTimer = setTimeout(async () => {
+      const result = await eventDomain.processNextEvent(runEvent);
+      respond({
+        type: "status",
+        currentIndex: eventDomain["eventIndex"],
+        timestamp: getFileNameForDate(eventDomain["cursor"]),
+        total: eventDomain["events"].length,
+      });
+      respond({
+        type: "result",
+        source: latestSource,
+        ...result,
+      });
+      queueNextEvent();
+    }, eventDelay);
+  });
+};
+
+self.addEventListener("message", async (event) => {
   const message = event.data as Request<string>;
 
-  if (message.type === "compile") {
+  if (message.type === "compile" || message.type === "init") {
     const { source } = message;
     latestSource = source;
-    debounceCompile();
-  } else if (message.type === "event") {
-    if (latestExecutable) {
-      runEvent(message.event, message.requestFeatures).then(
-        (props) => {
-          respond({
-            type: "result",
-            source: latestSource,
-            ...props,
-          });
-        },
-        (err) => {
-          respond({
-            type: "runtimeError",
-            message: err.stack,
-            source: latestSource,
-          });
-        }
-      );
+    if (message.type === "init") {
+      fetchFeatures = message.fetchFeatures;
+      clearTimeout(eventProcessTimer);
+      await triggerCompile();
+      eventDomain = new EventDomain(message);
+      queueNextEvent();
+    } else {
+      debounceCompile();
     }
+    return;
+  } else if (message.type === "playhead") {
+    clearTimeout(eventProcessTimer);
+    respond({
+      type: "status",
+      currentIndex: 0,
+      total: 0,
+      timestamp: getFileNameForDate(eventDomain["cursor"]),
+    });
+    eventDomain.setCursor(message.position);
+    queueNextEvent();
```

Comment:
We're probably going to want to clear the databases on seek, but can figure that out later.